Whitespace incorrectly suppressed inside table.

karlcow opened the pull request:
#### b57725d535269f4e209cfa4a08fddcb751ad2993
<pre>
Whitespace incorrectly suppressed inside table.
<a href="https://bugs.webkit.org/show_bug.cgi?id=251178">https://bugs.webkit.org/show_bug.cgi?id=251178</a>
<a href="https://rdar.apple.com/104934417">rdar://104934417</a>

Reviewed by NOBODY (OOPS!).

This fixes an issue where blink and webkit ignores the whitespace text
nodes in tables against what the specification is saying. See the issue
<a href="https://github.com/w3c/csswg-drafts/issues/8358">https://github.com/w3c/csswg-drafts/issues/8358</a>

The correct patch has been found by Ahmad Saleem.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063-expected.txt:
* LayoutTests/platform/mac/fast/css-generated-content/015-expected.txt:
* LayoutTests/platform/mac/fast/css-generated-content/table-before-after-child-add-expected.txt:
* LayoutTests/platform/mac/fast/table/table-after-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-before-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-cell-after-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-cell-before-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-row-after-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-row-before-child-in-table-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2479-2-expected.txt:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::textRendererIsNeeded):

Co-authored-by: Ahmad Saleem &lt;ahmad.saleem792+github@gmail.com&gt;
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b57725d535269f4e209cfa4a08fddcb751ad2993

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132934 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5435 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44026 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140469 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84965 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/66ffb7e9-05db-4d2b-a2f9-71c3640d9920) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134804 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5961 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5298 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101650 "Found 4 new test failures: accessibility/list-detection.html imported/w3c/web-platform-tests/css/css-tables/anonymous-table-ws-001.html imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter.html imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063.xhtml (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68980 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d67fce7e-9fe0-42d0-962f-be1a4f31d5dc) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135880 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4187 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119104 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82451 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4e15682f-5eb4-4f0a-b55c-61d5b8a4bb52) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4070 "Found 3 new test failures: imported/w3c/web-platform-tests/css/css-tables/anonymous-table-ws-001.html imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter.html imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063.xhtml (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1611 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83702 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113128 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37221 "Found 10 new test failures: fast/css-generated-content/015.html fast/css-generated-content/table-before-after-child-add.html fast/table/table-after-child-in-table.html fast/table/table-before-child-in-table.html fast/table/table-cell-after-child-in-table.html fast/table/table-cell-before-child-in-table.html fast/table/table-row-after-child-in-table.html fast/table/table-row-before-child-in-table.html imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter.html imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063.xhtml (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143122 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5104 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37804 "Found 10 new test failures: fast/css-generated-content/015.html fast/css-generated-content/table-before-after-child-add.html fast/table/table-after-child-in-table.html fast/table/table-before-child-in-table.html fast/table/table-cell-after-child-in-table.html fast/table/table-cell-before-child-in-table.html fast/table/table-row-after-child-in-table.html fast/table/table-row-before-child-in-table.html imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter.html imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063.xhtml (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110027 "Found 4 new test failures: accessibility/list-detection.html imported/w3c/web-platform-tests/css/css-tables/anonymous-table-ws-001.html imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter.html imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063.xhtml (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5186 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4370 "Found 10 new test failures: fast/css-generated-content/015.html fast/css-generated-content/table-before-after-child-add.html fast/table/table-after-child-in-table.html fast/table/table-before-child-in-table.html fast/table/table-cell-after-child-in-table.html fast/table/table-cell-before-child-in-table.html fast/table/table-row-after-child-in-table.html fast/table/table-row-before-child-in-table.html imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter.html imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063.xhtml (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110208 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3914 "Passed tests") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115367 "Exiting early after 10 failures. 19 tests run. 1 failures") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58669 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5158 "Built successfully") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33733 "Found 10 new test failures: fast/css-generated-content/015.html fast/css-generated-content/table-before-after-child-add.html fast/table/table-after-child-in-table.html fast/table/table-before-child-in-table.html fast/table/table-cell-after-child-in-table.html fast/table/table-cell-before-child-in-table.html fast/table/table-row-after-child-in-table.html fast/table/table-row-before-child-in-table.html imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter.html imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063.xhtml (failure)") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4998 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68610 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5248 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5116 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/30955398aa3aba9d15835b088f5ea3959ca2066e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47169

karlcow commented:
```
imported/blink/fast/table/whitespace-between-elements-with-table-cell-display-2.html
imported/blink/fast/table/whitespace-between-elements-with-table-cell-display-3.html
imported/blink/fast/table/whitespace-between-elements-with-table-cell-display.html
```

These 3 tests are being broken by this change, while Firefox is still behaving the same. The progression creates regression for form elements inside table elements. So something to figure out.

karlcow commented:
It also creates interesting regression for 
http://wpt.live/css/css-text/white-space/ws-break-spaces-applies-to-006.html

<img width="1725" alt="Screenshot 2025-06-26 at 07 14 33" src="https://github.com/user-attachments/assets/504d835a-7b4b-4739-8645-1e955703a266" />

karlcow commented:
```
imported/blink/fast/table/whitespace-between-elements-with-table-cell-display-3.html
```

<img width="1720" alt="Screenshot 2025-06-26 at 16 43 58" src="https://github.com/user-attachments/assets/abed7781-7299-4e8b-ab20-c15d0b01f5fe" />

karlcow committed:
Whitespace incorrectly suppressed inside table.
https://bugs.webkit.org/show_bug.cgi?id=251178
rdar://104934417

Reviewed by NOBODY (OOPS!).

This fixes an issue where blink and webkit ignores the whitespace text
nodes in tables against what the specification is saying. See the issue
https://github.com/w3c/csswg-drafts/issues/8358

The correct patch has been found by Ahmad Saleem.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/the-innertext-and-outertext-properties/getter-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-mozilla-162063-expected.txt:
* LayoutTests/platform/mac/fast/css-generated-content/015-expected.txt:
* LayoutTests/platform/mac/fast/css-generated-content/table-before-after-child-add-expected.txt:
* LayoutTests/platform/mac/fast/table/table-after-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-before-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-cell-after-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-cell-before-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-row-after-child-in-table-expected.txt:
* LayoutTests/platform/mac/fast/table/table-row-before-child-in-table-expected.txt:
* LayoutTests/platform/mac/tables/mozilla/bugs/bug2479-2-expected.txt:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::textRendererIsNeeded):

Co-authored-by: Ahmad Saleem <ahmad.saleem792+github@gmail.com>

Ahmad-S792 commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -570,12 +570,37 @@ bool RenderTreeUpdater::textRendererIsNeeded(const Text& textNode)
+
+        // Preserve whitespace only for inline-block and inline-table.
+        auto display = previousRenderer->style().display();
+        if (display == DisplayType::InlineBlock || display == DisplayType::InlineTable)
```
Just add `|| display == DisplayType::Inline` and it will fix failure where it is just `<span>` (inline layout).

Ahmad-S792 replied:
but it will fail - http://wpt.live/css/css-display/display-contents-dynamic-table-002-inline.html
